arch/risc-v/src/mpfs/mpfs_ethernet.c

```diff
@@ -0,0 +1,3860 @@
+  priv->queue[1].rx_q_ptr       = (uint32_t *)(base + RECEIVE_Q1_PTR);
+  priv->queue[2].rx_q_ptr       = (uint32_t *)(base + RECEIVE_Q2_PTR);
+  priv->queue[3].rx_q_ptr       = (uint32_t *)(base + RECEIVE_Q3_PTR);
+  priv->queue[0].tx_q_ptr       = (uint32_t *)(base + TRANSMIT_Q_PTR);
+  priv->queue[1].tx_q_ptr       = (uint32_t *)(base + TRANSMIT_Q1_PTR);
+  priv->queue[2].tx_q_ptr       = (uint32_t *)(base + TRANSMIT_Q2_PTR);
+  priv->queue[3].tx_q_ptr       = (uint32_t *)(base + TRANSMIT_Q3_PTR);
+  priv->queue[0].dma_rxbuf_size = (uint32_t *)(base + DMA_RXBUF_SIZE_Q1);
+  priv->queue[1].dma_rxbuf_size = (uint32_t *)(base + DMA_RXBUF_SIZE_Q1);
+  priv->queue[2].dma_rxbuf_size = (uint32_t *)(base + DMA_RXBUF_SIZE_Q2);
+  priv->queue[3].dma_rxbuf_size = (uint32_t *)(base + DMA_RXBUF_SIZE_Q3);
+
+  /* MPU hack for ETH DMA if not enabled by bootloader */
+
+#ifdef CONFIG_MPFS_MPU_DMA_ENABLE
+#  ifdef CONFIG_MPFS_ETHMAC_0
+  putreg64(0x1f00000fffffffff, MPFS_PMPCFG_ETH0_0);
+#  endif
+#  ifdef CONFIG_MPFS_ETHMAC_1
+  putreg64(0x1f00000fffffffff, MPFS_PMPCFG_ETH1_0);
+#  endif
+#endif
+
+  /* Allocate buffers */
+
+  ret = mpfs_buffer_initialize(priv, 0);
+  if (ret < 0)
+    {
+      nerr("ERROR: mpfs_buffer_initialize failed: %d\n", ret);
+      return ret;
+    }
+
+  /* Attach the IRQ to the driver */
+
+  if (irq_attach(priv->mac_q_int[0], mpfs_interrupt_0, priv))
+    {
+      /* We could not attach the ISR to the interrupt */
+
+      return -EAGAIN;
+    }
+
+  if (irq_attach(priv->mac_q_int[1], mpfs_interrupt_1, NULL))
+    {
+      /* We could not attach the ISR to the interrupt */
+
+      return -EAGAIN;
+    }
+
+  if (irq_attach(priv->mac_q_int[2], mpfs_interrupt_2, NULL))
+    {
+      /* We could not attach the ISR to the interrupt */
+
+      return -EAGAIN;
+    }
+
+  if (irq_attach(priv->mac_q_int[3], mpfs_interrupt_3, NULL))
+    {
+      /* We could not attach the ISR to the interrupt */
+
+      return -EAGAIN;
+    }
+
+  /* Enable clocking to the GMAC peripheral (just for mpfs_ifdown()) */
+
+  /* MAC HW clock enable and reset */
+
+  if (priv->regbase == MPFS_GEM0_LO_BASE)
+    {
+      modifyreg32(MPFS_SYSREG_BASE + MPFS_SYSREG_SUBBLK_CLOCK_CR_OFFSET, 0,
+                  SYSREG_SUBBLK_CLOCK_CR_MAC0);
+      modifyreg32(MPFS_SYSREG_BASE + MPFS_SYSREG_SOFT_RESET_CR_OFFSET,
+                  0, SYSREG_SOFT_RESET_CR_MAC0);
+
+      for (volatile int i = 0; i < 10000; i++)
+        {
+          ;
+        }
+
+      modifyreg32(MPFS_SYSREG_BASE + MPFS_SYSREG_SOFT_RESET_CR_OFFSET,
+                  SYSREG_SOFT_RESET_CR_MAC0, 0);
+    }
+
+  if (priv->regbase == MPFS_GEM1_LO_BASE)
+    {
+      modifyreg32(MPFS_SYSREG_BASE + MPFS_SYSREG_SUBBLK_CLOCK_CR_OFFSET, 0,
+                  SYSREG_SUBBLK_CLOCK_CR_MAC1);
+      modifyreg32(MPFS_SYSREG_BASE + MPFS_SYSREG_SOFT_RESET_CR_OFFSET,
+                  0, SYSREG_SOFT_RESET_CR_MAC1);
+
+      for (volatile int i = 0; i < 10000; i++)
+        {
+          ;
+        }
+
```

Comment:
can we use `up_udelay` instead?

Reply:
I'll remove this. it was attempt to get reset working on some weird cases. but has no effect actually.